## **Project. Neural Networks for Classification**
- Student ID: 

- Full name: 

### **Import Libraries**

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import tensorflow as tf
from tensorflow.keras import datasets, layers, models

### **1. Data Preparation**

#### **1.1 Data collection**

In [ ]:
(X_train, y_train), (X_test, y_test) = datasets.cifar10.load_data()
X_train.shape

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 21s 0us/step


(50000, 32, 32, 3)

In [30]:
y_train = y_train.reshape(-1,)
y_train

array([6, 9, 9, ..., 9, 1, 1], dtype=uint8)

In [37]:
classes = ["airplane", "automobile", "bird", "cat", "deer", "dog", "frog", "horse", "ship", "truck"]

#### **1.2 Data Nomalization**
It can be observed that the value of each pixel in an image is in the range of 0-255. We need to normalize it to the range (0, 1).

In [39]:
def nomalize(X):
    return X/255

X_train = nomalize(X_train)
X_test = nomalize(X_test)

In [42]:
(X_train.min(), X_train.max())

(0.0, 1.0)


Now our data has been normalized to the range (0, 1).

#### **1.3 Flatten**

In [43]:
X_train.shape

(50000, 32, 32, 3)


We can see that our image has dimensions of 32x32x3, representing an image with a size of 32x32 and 3 RGB channels. Now, we need to flatten it.

In [44]:
def flatten_images(images):
    return images.reshape(images.shape[0], -1)  # Flatten each image

X_train = flatten_images(X_train)
X_test = flatten_images(X_test)

In [46]:
X_train[0].shape

(3072,)

Now, our image data has been flattened, with the size of a single input being (32x32x3 = 3072)

#### **1.4 Encoding**

In [47]:
y_train

array([6, 9, 9, ..., 9, 1, 1], dtype=uint8)


Our output will consist of labels ranging from 0 to 9, corresponding to the classes: airplane, automobile, bird, cat, deer, dog, frog, horse, ship, and truck.

In [49]:
def one_hot_encode(labels, num_classes=10):
    return np.eye(num_classes)[labels]  # Convert to one-hot encoded array

y_train = one_hot_encode(y_train)
y_test = one_hot_encode(y_test)

In [50]:
y_train

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 0., 0., ..., 0., 0., 1.],
       ...,
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.]])


After encoding, each output will be a binary vector of length 10. The position with a value of 1 corresponds to the label. For example, if the 4th position has a value of 1, the label is 4, which corresponds to *deer*.